driver-core/src/main/java/com/datastax/driver/core/policies/RackAwareRoundRobinPolicy.java

```diff
@@ -0,0 +1,485 @@
+    return (CopyOnWriteArrayList<Host>) list.clone();
+  }
+
+  /**
+   * Return the HostDistance for the provided host.
+   *
+   * <p>This policy consider nodes in the local datacenter as {@code LOCAL}. For each remote
+   * datacenter, it considers a configurable number of hosts as {@code REMOTE} and the rest is
+   * {@code IGNORED}.
+   *
+   * <p>To configure how many hosts in each remote datacenter should be considered, see {@link
+   * Builder#withUsedHostsPerRemoteDc(int)}.
+   *
+   * @param host the host of which to return the distance of.
+   * @return the HostDistance to {@code host}.
+   */
+  @Override
+  public HostDistance distance(Host host) {
+    String dc = dc(host);
+    if (dc == UNSET || dc.equals(localDc)) return HostDistance.LOCAL;
+
+    CopyOnWriteArrayList<Host> dcHosts = perDcLiveHosts.get(dc);
+    if (dcHosts == null || usedHostsPerRemoteDc == 0) return HostDistance.IGNORED;
+
+    // We need to clone, otherwise our subList call is not thread safe
+    dcHosts = cloneList(dcHosts);
+    return dcHosts.subList(0, Math.min(dcHosts.size(), usedHostsPerRemoteDc)).contains(host)
+        ? HostDistance.REMOTE
+        : HostDistance.IGNORED;
+  }
+
+  /**
+   * Returns the hosts to use for a new query.
+   *
+   * <p>The returned plan will always try each known host in the local rack first, and then, if none
+   * of the host in local rack are reachable, will try hosts in remote racks of the local
+   * datacenter. Finally, it will try up to a configurable number of other host per remote
+   * datacenter. The order of the local node in the returned query plan will follow a Round-robin
+   * algorithm.
+   *
+   * @param loggedKeyspace the keyspace currently logged in on for this query.
+   * @param statement the query for which to build the plan.
+   * @return a new query plan, i.e. an iterator indicating which host to try first for querying,
+   *     which one to use as failover, etc...
+   */
+  @Override
+  public Iterator<Host> newQueryPlan(String loggedKeyspace, final Statement statement) {
+
+    CopyOnWriteArrayList<Host> localLiveHosts = perDcLiveHosts.get(localDc);
+    final List<Host> localDCLocalRackHosts = new ArrayList<Host>();
+    final List<Host> localDCRemoteRackHosts = new ArrayList<Host>();
+    for (Host host : localLiveHosts) {
+      if (rack(host).equals(localRack)) {
+        localDCLocalRackHosts.add(host);
+      } else {
+        localDCRemoteRackHosts.add(host);
+      }
+    }
+    final int startIdx = index.getAndIncrement();
+
+    return new AbstractIterator<Host>() {
+
+      private int idx = startIdx;
+      private int remainingLocalDCLocalRack = localDCLocalRackHosts.size();
+      private int remainingLocalDCRemoteRack = localDCRemoteRackHosts.size();
+
+      // For remote Dcs
+      private Iterator<String> remoteDcs;
+      private List<Host> currentDcHosts;
+      private int currentDcRemaining;
```

Comment:
I know this is the wording from DcAware policy, but it's really confusing for me - I'd very much prefer something like `currentRemoteDcHosts` and `currentRemoteDcRemaining`